### PR TITLE
add `as_str` method to more `str`-iterators

### PR DESCRIPTION
**Pull request by oli-obk**
In the spirit of [`Chars::as_str`](http://manishearth.github.io/rust-internals-docs/core/str/struct.Chars.html#method.as_str)

### CODE REVIEW

File: src/libcore/str/pattern.rs [hunk 1]
```diff
@@ -596,6 +607,13 @@ impl<'a, 'b> StrSearcher<'a, 'b> {
 unsafe impl<'a, 'b> Searcher<'a> for StrSearcher<'a, 'b> {
     fn haystack(&self) -> &'a str { self.haystack }
 
+    fn as_str(&self) -> &'a str {
+        match self.searcher {
+            StrSearcherImpl::Empty(ref searcher) => &self.haystack[searcher.position..searcher.end],
+            StrSearcherImpl::TwoWay(ref searcher) => &self.haystack[searcher.position..searcher.end],
```
Comment (kennytm) [inline src/libcore/str/pattern.rs]: Looks like this line is too long (> 100 chars) 😄

Comment (oli-obk) [inline src/libcore/str/pattern.rs]: fixed

**Comment by brson** [comment]
Floodgates are opened I guess. *grumble*.

**Comment by alexcrichton** [comment]
I think that this is a breaking change due to the addition of a non-default method on `Searcher`, right?

**Comment by alexcrichton** [comment]
Breaking change for nightly users, that is.

**Comment by oli-obk** [comment]
> Floodgates are opened I guess. grumble.

:/ I'm not depending on this, I just noticed the discrepancy. Feel free to close this PR and request an RFC if you think this should be properly designed.

An alternative would be to deprecate the `as_str` methods and create an RFC on how to properly get the remaining `str` or `[T]` or other ranged type for iterators (e.g. some might want to not drop the allocation in a `Vec::into_iterator`, but get it back).

> Breaking change for nightly users, that is.

indeed

**Comment by bluss** [comment]
There are forward *and* reverse iterators that are not double ended. Those correspond to two different string slices (one from the front cursor until the end, the other from the back cursor until the start)

**Comment by alexcrichton** [comment]
Discussed during libs triage yesterday, in light of @bluss's realization and lack fo strong motivation for this we're going to close at this time, but thanks though for the PR @oli-obk!

File: src/libcore/str/pattern.rs [hunk 2]
```diff
@@ -596,6 +607,17 @@ impl<'a, 'b> StrSearcher<'a, 'b> {
 unsafe impl<'a, 'b> Searcher<'a> for StrSearcher<'a, 'b> {
     fn haystack(&self) -> &'a str { self.haystack }
 
+    fn as_str(&self) -> &'a str {
+        match self.searcher {
+            StrSearcherImpl::Empty(ref searcher) => {
+                &self.haystack[searcher.position..searcher.end]
+            },
+            StrSearcherImpl::TwoWay(ref searcher) => {
+                &self.haystack[searcher.position..searcher.end]
```
Comment (bluss) [inline src/libcore/str/pattern.rs]: For example, position and end have no relation and position > end is a legal state.

**Pull request closed without merging**